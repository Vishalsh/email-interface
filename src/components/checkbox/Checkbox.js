import React from 'react';
import PropTypes from 'prop-types';

import classes from './Checkbox.module.scss';

const Checkbox = (props) => {
  const {
    labelClass,
    labelFor,
    onChange
  } = props;

  return (
    <>
    <input id={`checkbox_${labelFor}`}
           className={classes.input}
           type="checkbox"
           onChange={onChange}/>
    <label className={`${classes.label} ${labelClass}`}
           htmlFor={`checkbox_${labelFor}`}/>
    </>
  );
};

Checkbox.defaultProps = {
  labelClass: ''
};

Checkbox.propTypes = {
  labelClass: PropTypes.string,
  labelFor: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired
};

export default Checkbox;
