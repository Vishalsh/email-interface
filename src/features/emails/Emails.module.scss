@import "~styles/variables";

.emailList {
  max-height: calc(100vh - 19rem);
  overflow-y: scroll;
  overflow-x: hidden;
}