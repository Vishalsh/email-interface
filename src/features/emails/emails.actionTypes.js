export const ADD_EMAILS = 'ADD_EMAILS';
