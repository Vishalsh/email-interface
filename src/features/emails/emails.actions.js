import { createAction } from 'redux-actions';

import {
  ADD_EMAILS,
} from "./emails.actionTypes";

const addEmails = createAction(ADD_EMAILS);

export default {
  addEmails
};
