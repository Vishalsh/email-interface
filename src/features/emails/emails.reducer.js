import {
  ADD_EMAILS,
} from "./emails.actionTypes";

const initialState = {};

const emailsReducer = (state = initialState, action) => {
  switch (action.type) {
    case ADD_EMAILS: {
      const { emails } = action.payload;

      return {
        ...state,
        ...emails
      };
    }
    default:
      return state;
  }
};

export default emailsReducer;
