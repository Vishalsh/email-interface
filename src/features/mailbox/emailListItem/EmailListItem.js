import React from 'react';
import PropTypes from 'prop-types';

import classes from './EmailListItem.module.scss';

const emailStatus = {
  READ: 'READ',
  UNREAD: 'UNREAD'
};

const EmailListItem = (props) => {
  const {
    email
  } = props;

  return (
    <li className={classes.listItem}>
      <article className={`row middle-xs ${classes.emailRow} ${email.status === emailStatus.UNREAD ? classes.unread : ''}`}>
        <span className="col-sm-2">{email.sender.name}</span>
        <p className="col-sm-2 end-sm">
          <span className={`${classes.category} ${classes[email.category]}`}>
            {email.category}
          </span>
        </p>
        <p className="col-sm-5">{email.body}</p>
        <span className="col-sm-1 center-sm">
          {
            email.attachments && email.attachments.length > 0 &&
            <i className={`icon icon-attachment ${classes.icon}`}/>
          }
        </span>
        <span className="col-sm-2 end-sm">{email.dateTime}</span>
      </article>
    </li>
  );
};

EmailListItem.propTypes = {
  email: PropTypes.object.isRequired
};

export default EmailListItem;
