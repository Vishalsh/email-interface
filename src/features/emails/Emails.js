import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import EmailListItem from 'features/emails/emailListItem/EmailListItem';
import classes from './Emails.module.scss';

export const Emails = (props) => {
  const {
    emails
  } = props;

  return (
    <ul className={classes.emailList}>
      {
        emails.map((email) => (
          <EmailListItem key={email.id} email={email}/>
        ))
      }
    </ul>
  );
};

Emails.propTypes = {
  emails: PropTypes.array.isRequired
};

const mapStateToProps = (store, ownProps) => ({
  emails: Object.keys(store.emails).length > 0 ?
    ownProps.ids.map(id => store.emails[id]) : []
});

export default connect(mapStateToProps)(Emails);
