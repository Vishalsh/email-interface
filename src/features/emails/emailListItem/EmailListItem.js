import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import routes from 'constants/routes';

import { status } from 'constants/emails';
import Badge from 'components/badge/Badge';
import classes from './EmailListItem.module.scss';

const EmailListItem = (props) => {
  const {
    mailbox,
    email,
    onClickEmail
  } = props;

  const onClickLink = () => {
    if (email.status === status.UNREAD) {
      onClickEmail(email);
    }
  };

  return (
    <li className={classes.listItem}>
      <Link className={`row middle-xs ${classes.emailRow} ${email.status === status.UNREAD ? classes.unread : ''}`}
            to={`${routes.MAILBOX}/${mailbox}/${email.id}`}
            onClick={onClickLink}>
        <span className="col-sm-2">{email.sender.name}</span>
        <p className="col-sm-2 end-sm">
          <span className={classes.category}>
            <Badge className={email.category} label={email.category}/>
          </span>
        </p>
        <p className="col-sm-5">{email.subject}</p>
        <span className="col-sm-1 center-sm">
          {
            email.attachments && email.attachments.length > 0 &&
            <i className={`icon icon-attachment ${classes.icon}`}/>
          }
        </span>
        <span className="col-sm-2 end-sm">{email.dateTime}</span>
      </Link>
    </li>
  );
};

EmailListItem.propTypes = {
  email: PropTypes.object.isRequired,
  mailbox: PropTypes.string.isRequired,
  onClickEmail: PropTypes.func.isRequired
};

export default EmailListItem;
