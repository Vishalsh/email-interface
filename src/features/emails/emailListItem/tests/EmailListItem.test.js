import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Link } from 'react-router-dom';

import { status } from 'constants/emails';
import Checkbox from 'components/checkbox/Checkbox';
import EmailListItem from '../EmailListItem';

describe('EmailListItem', () => {
  const props = {
    email: {
      id: 1,
      subject: 'subject 1',
      sender: {
        name: 'MS Dhoni',
        email: 'msDhoni@bcci.com'
      },
      attachments: ['attachment 1', 'attachment 1'],
      dateTime: '15 Jan',
      category: 'documents',
      status: status.READ
    },
    mailbox: 'inbox',
    onClickEmail: jest.fn(),
    onChangeCheckbox: jest.fn()
  };
  const component = shallow(<EmailListItem {...props} />);

  it('should render the component', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should not trigger onClickEmail if the email status is not UNREAD', () => {
    component.find(Link).props().onClick();

    expect(props.onClickEmail).not.toHaveBeenCalled();
  });

  it('should not trigger onClickEmail if the email status is not UNREAD', () => {
    const propsWithUnreadEmail = {
      ...props,
      email: {
        ...props.email,
        status: status.UNREAD
      }
    };
    const componentWithUnreadEmail = shallow(<EmailListItem {...propsWithUnreadEmail} />);

    componentWithUnreadEmail.find(Link).props().onClick();

    expect(props.onClickEmail).toHaveBeenCalledWith(propsWithUnreadEmail.email);
  });

  it('should trigger onChangeCheckbox', () => {
    component.find(Checkbox).props().onChange({ target: { checked: true } });

    expect(props.onChangeCheckbox).toHaveBeenCalledWith({ id: props.email.id, checked: true });
  });
});