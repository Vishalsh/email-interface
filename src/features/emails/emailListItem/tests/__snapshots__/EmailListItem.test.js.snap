// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailListItem should render the component 1`] = `
<li
  className="listItem"
>
  <article
    className="row middle-xs emailRow "
  >
    <span
      className="col-sm-2"
    >
      MS Dhoni
    </span>
    <p
      className="col-sm-2 end-sm"
    >
      <span
        className="category undefined"
      />
    </p>
    <p
      className="col-sm-5"
    >
      subject 1
    </p>
    <span
      className="col-sm-1 center-sm"
    >
      <i
        className="icon icon-attachment icon"
      />
    </span>
    <span
      className="col-sm-2 end-sm"
    >
      15 Jan
    </span>
  </article>
</li>
`;
